File: javascript/src/lib/factors.js
/**
 * Iterates over the proper divisors of a given number
 * @param {number} num
 * @yield {number}
 */
function* properDivisors(num) {
    factors = [...primes.primeFactors(num)];
    const seen = new Set();
    yield 1;
    for (let x = 1; x < factors.length; x++) {
        for (const combo of iters.combinations(factors, x)) {
            ret = combo.reduce((a, x) => a * x, 1);
            if (!seen.has(ret)) {
                yield ret;
                seen.add(ret);
            }
        }
        seen.clear();
    }
}
exports.properDivisors = properDivisors;

const iters = require('./iters.js');
const primes = require('./primes.js');

File: python/src/lib/factors.py
from functools import reduce
from itertools import combinations
from operator import mul
from typing import Iterator, Set

from .primes import prime_factors


def proper_divisors(num: int) -> Iterator[int]:
    """Iterate over the proper divisors of a given number.
    
    Given that it will generate a max of :math:`\\log(n)` factors, the binomial theorem tells us this should take
    :math:`O(2^{\\log(n)}) = O(n)` operations."""
    factors = tuple(prime_factors(num))
    seen: Set[int] = set()
    yield 1
    for x in range(1, len(factors)):
        for combo in combinations(factors, x):
            ret = reduce(mul, combo, 1)
            if ret not in seen:
                yield ret
                seen.add(ret)
        seen.clear()
